presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -498,6 +498,13 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                 }
             }
 
+            if (node.getDistributionType().isPresent()
+                    && JoinNode.requiresEquiJoinClauses(node.getDistributionType().get(), node.getType())
+                    && equiJoinClauses.isEmpty()) {
+                verify(!node.getCriteria().isEmpty(), "Original criteria was empty");
+                equiJoinClauses.add(node.getCriteria().get(0));
+            }
```

Comment:
Please add a comment explaining why we need to do this.

Reply:
I structured the condition so that i call `requiresEquiJoinClauses` (new method) and this links this safety `if` to the check in JoinNode ctor.
so i thought no more comment is needed. can you suggest something?

Reply:
I was thinking something that explains *why* equijoin clauses might be needed and why we even need this safety check (in part, so that we know when it's safe to remove in the future) -- the how and what is obvious from reading the code, but the underlying motivation is not.

Reply:
Then this boils down to why exactly we have the check in JoinNode constructor, that we talked about.
However _I_ do not know exactly why.
(I removed both: the fix and ctor check and my test passed in local & distributed manner, but that doesn't prove anything.)

Reply:
copying @sopel39 into this thread

> it seems that this check is unnecessary really. It's sub-optimal to use partitioned distribution without any equi-clause, but the execution shouldn't fail (partitioning on empty symbols is constant).
> 
> Maybe we should remove the check here? We could have a plan sanity checker instead or generate warning in such case.

This is exactly what I proposed initially on Slack, so I tend to agree.

I tried to find where it would fail

- i changed ReorderJoins (because it's run in appropriate moment), to always choose PARTITIONED, no matter what:

```diff
--- presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ReorderJoins.java
+++ presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ReorderJoins.java
@@ -129,6 +129,10 @@ public class ReorderJoins
     @Override
     public Result apply(JoinNode joinNode, Captures captures, Context context)
     {
+        if (Boolean.TRUE) {
+            return Result.ofPlanNode(joinNode.withDistributionType(PARTITIONED));
+        }
+
         // try reorder joins with projection pushdown first
```

- this made 37 of 113 tests in `TestHiveDistributedJoinQueriesWithoutDynamicFiltering` fail in JoinNode ctor

- i removed the check from JoinNode ctor

```
--- presto-main/src/main/java/io/prestosql/sql/planner/plan/JoinNode.java
+++ presto-main/src/main/java/io/prestosql/sql/planner/plan/JoinNode.java
@@ -136,12 +136,6 @@ public class JoinNode
                     "%s join do not work with %s distribution type",
                     type,
                     distributionType.get());
-            // It does not make sense to PARTITION when there is nothing to partition on
-            checkArgument(
-                    !(distributionType.get() == PARTITIONED && criteria.isEmpty() && type != RIGHT && type != FULL),
-                    "Equi criteria are empty, so %s join should not have %s distribution type",
-                    type,
-                    distributionType.get());
         }
```

- and `TestHiveDistributedJoinQueriesWithoutDynamicFiltering` passed

What else should I check?

Reply:
AddExchanges relies on the distribution type and equi-join criteria when calculating preferred and required properties. Having a partitioned join with no equijoin criteria should be messing that up (possibly introducing unnecessary exchanges)

Reply:
https://github.com/prestosql/presto/pull/3245

if we keep the artificial equi condition, we will surely have  unnecessary exchanges too

Reply:
Not if there's something that might depend on or be able to leverage the same partitioning criteria above or below the join.

Reply:
If the exchanges are already added, they already captured the relevant partitioning.

if the exchanges are not yet added --
In the case where this matters, we are converting partitioned join to a cross join because join condition was removed, pushed down.
So, this partitioned join actually ALREADY was single-threaded (in disguise), as there is only one value for the join key on both side.
From PPD point of view we have options

1. keep at least one equi condition (this PR, already approved), faking the cross join to look as partitioned join
2. remove the equi condition (https://github.com/prestosql/presto/pull/3245), making it clear that this is single threaded join

Furthermore, we can add a new PPD run between CBO/RJ and AddExchanges to push down DFs and configure this run to allow it to revert RJ's decisions about distribution type. I can work on this. This is a good improvement to do.
But it doesn't fix the underlying issue -- the bug where PPD can blow up violating JoinNode ctor contraints, which brings as back to 1. or 2. -- So I think we should merge 1. or 2. 

@sopel39 @martint Am I missing something? Any other options?

Reply:
I think we might be overcomplicating things.
We know that PPD might simplify join after partitioning has already been set based on new predicates.  Keep not that the join was already inefficient even though it had predicate.
We can simply relax `JoinNode` condition. It won't break query or make it any worse.